### PR TITLE
Explicitly specify pad token id when generating tokens 

### LINKED ISSUE #3536: Explicitly specify pad token id when generating tokens
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
We have added a [fallback pad token if it is not already present in the tokenizer](https://github.com/kserve/kserve/blob/11e0ab2ca4d852ec154156fad4f9d220c8768d78/python/huggingfaceserver/huggingfaceserver/model.py#L130-L131) as part of the PR #3459. But it does not explicitly specifies pad_token_id when invoking generate method which leads to huggingface using eos_token_id as the pad_token_id. The log from huggingface server is show below. To avoid this we should explicitly specify the pad_token_id
```
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
```

**What did you expect to happen:**
pad_token_id should be set accordingly.

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]
```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: huggingface-gpt2
spec:
  predictor:
    model:
      modelFormat:
        name: huggingface
      args:
        - "--model_name=gpt2"
        - "--model_id=openai-community/gpt2"
        - "--tensor_input_names=input_ids"
        - "--disable_vllm"
        - "--task=6"
      env:
        - name: CUDA_LAUNCH_BLOCKING
          value: "1"
      resources:
        limits:
          nvidia.com/gpu: "1"
        requests:
          nvidia.com/gpu: "1"
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version: master
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:
We have added a [fallback pad token if it is not already present in the tokenizer](https://github.com/kserve/kserve/blob/11e0ab2ca4d852ec154156fad4f9d220c8768d78/python/huggingfaceserver/huggingfaceserver/model.py#L130-L131) as part of the PR #3459. But it does not explicitly specifies pad_token_id when invoking generate method which leads to huggingface using eos_token_id as the pad_token_id. The log from huggingface server is show below. To avoid this we should explicitly specify the pad_token_id
```
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
```
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3536

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Explicitly specify pad token id when generating tokens
```

### CODE REVIEW

**Comment by sivanantha321** [comment]
supersedes #3535

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 1]
```diff
@@ -198,6 +197,16 @@ def load(self) -> bool:
                 raise ValueError(
                     f"Unsupported task {self.task}. Please check the supported `task` option."
                 )
+            if not self.tokenizer.pad_token:
```
Comment (yuzisun) [inline python/huggingfaceserver/huggingfaceserver/model.py]: Seems like this is moved to only for the predictor case, for the transformer mode do we still need to apply the padding ?

Comment (sivanantha321) [reply python/huggingfaceserver/huggingfaceserver/model.py]: We won't have access to the model to update the vocabulary size. So, Even if we add the pad token we will get an `Index out of range ` error.

File: python/huggingfaceserver/huggingfaceserver/test_model.py [hunk 2]
```diff
@@ -175,5 +175,28 @@ def test_input_truncation():
...
+    # inputs with different lengths will throw an error
+    # unless padding token is configured.
+    # openai-gpt model does not specify the pad token, so the fallback pad token should be added.
+    assert model.tokenizer.pad_token == "[PAD]"
+    assert model.tokenizer.pad_token_id is not None
+    request_one = "my name is merve and my favorite color is blue."
+    request_two = "my name is teven and i am"
+    response = asyncio.run(model({"instances": [request_one, request_two]}, headers={}))
+    assert request_one in response["predictions"][0]
+    assert request_two in response["predictions"][1]
```
Comment (sivanantha321) [inline python/huggingfaceserver/huggingfaceserver/test_model.py]: I am asserting it this way because the output changes occasionally.

Comment (yuzisun) [reply python/huggingfaceserver/huggingfaceserver/test_model.py]: I think we can specify the temperature to 0 to get deterministic response

**Comment by saileshd1402** [comment]
Greetings!
There was a certain bug I faced while I was testing master branch which might be related to this PR:

CMD to run hf backend run:
```
python -m huggingfaceserver --model_id=microsoft/phi-2 --model_name=phi --backend=huggingface
```

cURL request:
```
curl -v -H "Content-Type: application/json" http://localhost:8080/openai/v1/completions -d '{"model":"phi", "prompt":"Hello give me a hello world python program", "stream": true, "max_tokens": 5}'
```

Error:
```
Exception in thread Thread-4 (generate):
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.11/threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ubuntu/.testenv/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.testenv/lib/python3.11/site-packages/transformers/generation/utils.py", line 1527, in generate
    result = self._greedy_search(
             ^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.testenv/lib/python3.11/site-packages/transformers/generation/utils.py", line 2452, in _greedy_search
    raise ValueError("If `eos_token_id` is defined, make sure that `pad_token_id` is defined.")
ValueError: If `eos_token_id` is defined, make sure that `pad_token_id` is defined.
```

After displaying this error, the runtime freezes and won't take any other requests

CC: @johnugeorge

**Comment by yuzisun** [comment]
@sivanantha321 can you help resolve the merge conflict ?

**Review by spolti — APPROVED** [review]
/lgtm

**Pull request merged**